Fix Dynamic Valve HashMap leak: Use instanceof in FluidRenderData.equals

LuminaSapphira opened the pull request:
Currently, each frame a Dynamic Valve is being rendered a new instance of a `Float2ObjectMap<Model3D>` is created and stored in `ModelRenderer.cachedValveFluids`. Although the code is written to use `computeIfAbsent`, the equality check for `ValveRenderData` always returns `false`, because its superclass, `FluidRenderData`, tests for exact class equality. By changing this check to an `instanceof` we can allow the `super.equals()` to succeed, preventing the map from growing infinitely and allowing the cache to function as intended.

Because the map can grow to quite a few objects, the lookup and insertion time can increase drastically, severely hurting performance in long play sessions. I've seen `ModelRenderer.getValveModel` take up to 80% of CPU time when profiled (recorded via Spark's client profiler). In the amount of time its taken to write the description of this PR, VisualVM's heap dump displays over 80k instances of `ValveRenderData` from looking at a single Dynamic Tank slowly filling with water on a test world:
<img width="508" height="284" alt="image" src="https://github.com/user-attachments/assets/ffd59a9c-70fd-4606-8466-1a8402d33912" />

After updating the `equals` implementation:

<img width="526" height="310" alt="image" src="https://github.com/user-attachments/assets/a64e7d7f-e836-41e5-8525-6d4230f93009" />

The count continues to stay fixed at 1 even when letting run for a long time - cache is working :)

LuminaSapphira committed:
Use instanceof in FluidRenderData.equals

thiakil reviewed (changes requested):
This is incorrect. Looks like it was missed in https://github.com/mekanism/Mekanism/commit/3aa83b0a5392cd9516ec04eb7db611ab265afcb8 and valve render data should use that helper method to check for equality

LuminaSapphira committed:
Revert "Use instanceof in FluidRenderData.equals"

This reverts commit d80b717c9695497cfc5216ad87dce4a16af28781.

LuminaSapphira commented:
Thank you for your prompt review! Are you referring to `equalsCommonFluid`? If so, wouldn't the `super.equals` call handle that?

For now, I've reverted the commit, and added a new commit with `getClass` comparison instead of `instanceof` - I've been out of touch with Java for a long while and forgot instanceof in `equals` was bad. This new class comparison calls `getClass` on the `this` instance as well instead of the static `FluidRenderData.class` to support children calling super.equals.

Lmk what changes you'd like to see in regards to using `super.equals` vs calling the helper method directly, and if you'd like a rebase to drop the reverted and revert commits.

LuminaSapphira committed:
Use getClass() instead of static .class
